eras/babbage/impl/src/Cardano/Ledger/Babbage/Scripts.hs

```diff
@@ -67,39 +69,60 @@ getDatum tx (UTxO m) sp = do
+babbageTxScripts ::
   forall era.
   ( Core.TxBody era ~ TxBody era,
-    UsesTxOut era,
+    Core.TxOut era ~ TxOut era,
     ValidateScript era
   ) =>
   UTxO era ->
   Core.Tx era ->
   Map.Map (ScriptHash (Crypto era)) (Core.Script era)
-txscripts (UTxO mp) tx = SplitMap.foldl' accum (getField @"scriptWits" tx) smallUtxo
+babbageTxScripts (UTxO mp) tx = SplitMap.foldl' accum (getField @"scriptWits" tx) smallUtxo
   where
     txbody = getField @"body" tx
     scriptinputs = referenceInputs' txbody `Set.union` spendInputs' txbody
     smallUtxo = scriptinputs SplitMap.◁ mp
     accum ans txout =
-      case txOutView @era txout of
-        (_, _, _, SNothing) -> ans
-        (_, _, _, SJust script) -> Map.insert (hashScript @era script) script ans
+      case txout of
+        (TxOut _ _ _ SNothing) -> ans
+        (TxOut _ _ _ (SJust script)) -> Map.insert (hashScript @era script) script ans
 
-languages ::
+-- Compute two sets for all TwoPhase scripts in a Tx.
+-- set 1) DataHashes for each Two phase Script in a TxIn that has a DataHash
+-- set 2) TxIns that are TwoPhase scripts, and should have a DataHash but don't.
+--        in Babbage, a TxOut with an inline Datum, does not need DataHash, so
+--        it should not be added to set of Bad TxIn.
+{- { h | (_ → (a,_,h)) ∈ txins tx ◁ utxo, isNonNativeScriptAddress tx a} -}
+babbageInputDataHashes ::
   forall era.
-  ( ValidateScript era,
-    UsesTxOut era,
-    Core.TxBody era ~ TxBody era,
-    Core.Script era ~ Script era
+  ( HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    ValidateScript era,
+    Core.TxOut era ~ TxOut era
   ) =>
+  Map.Map (ScriptHash (Crypto era)) (Core.Script era) ->
+  ValidatedTx era ->
   UTxO era ->
-  Core.Tx era ->
-  Set Language
-languages utxo tx =
-  Set.fromList
-    [ lang
-      | (_, script) <- Map.toList (txscripts utxo tx),
-        (not . isNativeScript @era) script,
-        Just lang <- [language @era script]
-    ]
+  (Set (DataHash (Crypto era)), Set (TxIn (Crypto era)))
+babbageInputDataHashes hashScriptMap tx (UTxO mp) =
+  SplitMap.foldlWithKey' accum (Set.empty, Set.empty) smallUtxo
+  where
+    txbody = body tx
+    spendinputs = getField @"inputs" txbody :: (Set (TxIn (Crypto era)))
+    smallUtxo = spendinputs SplitMap.◁ mp
+    accum ans@(hashSet, inputSet) txin txout =
+      case txout of
+        (TxOut addr _ NoDatum _) ->
+          if isTwoPhaseScriptAddressFromMap @era hashScriptMap addr
+            then (hashSet, Set.insert txin inputSet)
+            else ans
+        (TxOut addr _ (DatumHash dhash) _) ->
+          if isTwoPhaseScriptAddressFromMap @era hashScriptMap addr
+            then (Set.insert dhash hashSet, inputSet)
+            else ans
+        (TxOut addr _ (Datum _) _) ->
+          if isTwoPhaseScriptAddressFromMap @era hashScriptMap addr
+            then ans -- An a TwoPhaseScript with Explict Datum and does not need a DataHash
+            else (hashSet, Set.insert txin inputSet)
+
```

Comment:
What did we decide here?  Is a phase 1 script allowed to have an explicit datum.  If that is so then it should be
`   (TxOut addr _ (Datum _) _) -> ans`

As both of the branches of the if, need to do nothing.

Reply:
ah yes, good point. I think we wish to allow a phase 1 script allowed to have an explicit datum. I will change it and make a ticket/todo to confirm the choice.